<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="rover">

  <xacro:property name="base_width" value="0.6357" />
  <xacro:property name="base_len" value="0.563" />
  <xacro:property name="base_height" value="0.305" />
  <xacro:property name="wheel_radius" value="0.035" />
  <xacro:property name="base_wheel_gap" value="0.007" />
  <xacro:property name="wheel_separation" value="0.15" />
  <xacro:property name="wheel_joint_offset" value="0.02" />

  <xacro:property name="caster_wheel_radius" value="${wheel_radius/2}" />
  <xacro:property name="caster_wheel_mass" value="0.001" />
  <xacro:property name="caster_wheel_joint_offset" value="-0.052" />
  <!-- <xacro:property name="mesh_scale" value="0.0254 0.0254 0.0254" /> -->
  <xacro:property name="PI" value="3.141592" />

  <!--Color Properties-->
  <material name="blue">
    <color rgba="0 0 0.8 1" />
  </material>
  <material name="black">
    <color rgba="0 0 0 1" />
  </material>
  <material name="white">
    <color rgba="1 1 1 1" />
  </material>
  <material name="red">
    <color rgba="0.8 0.0 0.0 1.0" />
  </material>

  <!--Interial macros-->
  <xacro:macro name="cylinder_inertia" params="m r h">
    <inertial>
      <mass value="${m}" />
      <inertia ixx="${m*(3*r*r+h*h)/12}" ixy="0" ixz="0" iyy="${m*(3*r*r+h*h)/12}" iyz="0" izz="${m*r*r/2}" />
    </inertial>
  </xacro:macro>

  <xacro:macro name="box_inertia" params="m w h d">
    <inertial>
      <mass value="${m}" />
      <inertia ixx="${m / 12.0 * (d*d + h*h)}" ixy="0.0" ixz="0.0" iyy="${m / 12.0 * (w*w + h*h)}" iyz="0.0" izz="${m / 12.0 * (w*w + d*d)}" />
    </inertial>
  </xacro:macro>

  <xacro:macro name="sphere_inertia" params="m r">
    <inertial>
      <mass value="${m}" />
      <inertia ixx="${2.0*m*(r*r)/5.0}" ixy="0.0" ixz="0.0" iyy="${2.0*m*(r*r)/5.0}" iyz="0.0" izz="${2.0*m*(r*r)/5.0}" />
    </inertial>
  </xacro:macro>

  <!--Base Footprint-->
  <link name="base_footprint">
    <xacro:box_inertia m="10" w="0.001" h="0.001" d="0.001" />
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.001 0.001 0.001" />
      </geometry>
    </visual>
  </link>

  <!--Base link-->
  <link name="base_link">
    <xacro:box_inertia m="45.0" w="${base_len}" h="${base_width}" d="${base_height}" />
    <visual>
      <origin rpy="0 0 0" xyz="0 0 -0.03"/>
      <geometry>
        <mesh filename="package://mrover/rover_description/meshes/chassis.stl" scale="1 1 1"/>
      </geometry>
      <material name="blue" />
    </visual>
    <collision>
      <geometry>
        <box size="${base_len} ${base_width} ${base_height}" />
      </geometry>
    </collision>
  </link>

  <link name="right_rocker_link">
    <xacro:box_inertia m="3.0" w="${base_len}" h="${base_width}" d="${base_height}" />
    <visual>
      <origin rpy="0 0 0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://mrover/rover_description/meshes/right_rocker.stl" />
      </geometry>
      <material name="red" />
    </visual>
    <collision>
      <geometry>
        <box size="${base_len + 0.17} 0.05 ${base_height + 0.2}" />
      </geometry>
    </collision>
  </link>

  <joint name="base_right_rocker_joint" type="fixed">
    <origin xyz="0.01 ${-base_width/2.0 - 0.05} -0.2" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="right_rocker_link" />
  </joint>

  <link name="left_rocker_link">
    <xacro:box_inertia m="3.0" w="${base_len}" h="${base_width}" d="${base_height}" />
    <visual>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://mrover/rover_description/meshes/left_rocker.stl"/>
      </geometry>
      <material name="white" />
    </visual>
    <collision>
      <geometry>
        <box size="${base_len + 0.17} 0.05 ${base_height + 0.2}" />
      </geometry>
    </collision>
  </link>

  <joint name="base_left_rocker_joint" type="fixed">
    <origin xyz="0.01 ${base_width/2.0 + 0.05} -0.2" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="left_rocker_link" />
  </joint>

  <link name="left_bogie_link">
    <xacro:box_inertia m="3.0" w="${base_len}" h="${base_width}" d="${base_height}" />
    <visual>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://mrover/rover_description/meshes/left_bogie.stl"/>
      </geometry>
      <material name="white" />
    </visual>
    <collision>
      <geometry>
        <box size="${base_len + 0.17} 0.05 ${base_height + 0.2}" />
      </geometry>
    </collision>
  </link>

  <joint name="base_left_bogie_joint" type="fixed">
    <origin xyz="0.01 ${base_width/2.0 + 0.05} -0.2" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="left_bogie_link" />
  </joint>

  <link name="right_bogie_link">
    <xacro:box_inertia m="3.0" w="${base_len}" h="${base_width}" d="${base_height}" />
    <visual>
      <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://mrover/rover_description/meshes/right_bogie.stl"/>
      </geometry>
      <material name="white" />
    </visual>
    <collision>
      <geometry>
        <box size="${base_len + 0.17} 0.05 ${base_height + 0.2}" />
      </geometry>
    </collision>
  </link>

  <joint name="base_rught_bogie_joint" type="fixed">
    <origin xyz="0.01 ${-base_width/2.0 - 0.05} -0.2" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="right_bogie_link" />
  </joint>

  <!--base_link to base_footprint Joint-->
  <joint name="base_link_joint" type="fixed">
    <origin xyz="0 0 ${wheel_radius + 0.005}" rpy="0 0 0" />
    <parent link="base_footprint" />
    <child link="base_link" />
  </joint>

  <!--Wheel link & joint macro-->
  <xacro:macro name="wheel" params="prefix reflect">
    <link name="${prefix}_wheel">
      <visual>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
        <geometry>
          <cylinder radius="${wheel_radius}" length="0.005" />
        </geometry>
        <material name="red" />
      </visual>
      <collision>
        <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
        <geometry>
          <cylinder radius="${wheel_radius}" length="0.005" />
        </geometry>
      </collision>
      <xacro:cylinder_inertia m="10" r="${wheel_radius}" h="0.005" />
    </link>

    <joint name="${prefix}_wheel_joint" type="continuous">
      <axis xyz="0 1 0" rpy="0 0 0" />
      <parent link="base_link" />
      <child link="${prefix}_wheel" />
      <origin xyz="${wheel_joint_offset} ${((base_width/2)+base_wheel_gap)*reflect} -0.005" rpy="0 0 0" />
    </joint>
  </xacro:macro>

  <!--Create Left & Right Wheel links/joints-->
  <xacro:wheel prefix="left" reflect="1" />
  <xacro:wheel prefix="right" reflect="-1" />

  <!--Caster Wheel Link-->
  <link name="caster_wheel">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <sphere radius="${caster_wheel_radius}" />
      </geometry>
      <material name="blue" />
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <sphere radius="${caster_wheel_radius}" />
      </geometry>
    </collision>
    <xacro:sphere_inertia m="10" r="${caster_wheel_radius}" />
  </link>

  <!--Caster Wheel Joint-->
  <joint name="caster_wheel_joint" type="continuous">
    <axis xyz="0 1 0" rpy="0 0 0" />
    <parent link="base_link" />
    <child link="caster_wheel" />
    <origin xyz="${caster_wheel_joint_offset} 0 -${caster_wheel_radius+0.005}" rpy="0 0 0" />
  </joint>

  <xacro:include filename="$(find mrover)/rover_description/rover_gazebo_plugins.xacro" />

</robot>